File: CustomApplications/VB.Net/VGTTutorial/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports System.Data.Common
Imports System.Collections
Imports System.Collections.Specialized

Imports AGI.STKUtil
Imports AGI.STKX
Imports AGI.STKObjects
Imports AGI.STKVgt

Public Class VGTTutorial

#Region "Private Data Members"
    Private stkRootObject As AgStkObjectRoot = Nothing
    Private m_satellite As AgSatellite = Nothing
    Private m_facility As IAgFacility = Nothing
    Const m_sViewPointName As String = "ViewPoint"
    Const m_sDistanceVectorName As String = "accessFac"
#End Region

    Private ReadOnly Property stkRoot() As AGI.STKObjects.AgStkObjectRoot
        Get
            If stkRootObject Is Nothing Then
                stkRootObject = New AGI.STKObjects.AgStkObjectRoot()
                AddHandler stkRootObject.OnAnimUpdate, AddressOf stkRootObject_OnAnimUpdate
            End If
            Return stkRootObject
        End Get
    End Property

    Private Sub stkRootObject_OnAnimUpdate(ByVal TimeEpSec As Double)
        If m_satellite IsNot Nothing Then
            Dim provider As IAgCrdnProvider = m_satellite.Vgt
            If provider.Vectors.Contains(m_sDistanceVectorName) Then
                Dim vector As IAgCrdnVector = provider.Vectors(m_sDistanceVectorName)
                Dim result As IAgCrdnVectorFindInAxesResult = vector.FindInAxes(TimeEpSec, provider.WellKnownAxes.Earth.Fixed)
                If result.IsValid Then
                    ' Calculate the vector's L2-norm.
                    Dim cartVec As IAgCartesian3Vector = result.Vector
                    Dim mag As Double = Math.Sqrt(cartVec.X * cartVec.X + cartVec.Y * cartVec.Y + cartVec.Z * cartVec.Z)
                    TextBox1.Text = String.Format("{0} km", (mag / 1000).ToString("F"))
                End If
            End If
        End If
    End Sub
    Private Sub CloseScenario()
        If stkRootObject IsNot Nothing Then
            stkRootObject.CloseScenario()
        End If
    End Sub

    Public Sub New()
        InitializeComponent()
    End Sub

    Private Sub VGTTutorial_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        MainLabel.Text = My.Resources.Strings.Welcome
    End Sub
    Private Sub CreateScenario_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles CreateScenario.Click
        '--Set Instructions--
        MainLabel.Text = My.Resources.Strings.CreateSatellite

        '--Creates new scenario--
        stkRoot.NewScenario("VGTTutorial")
        stkRoot.UnitPreferences.SetCurrentUnit("TimeUnit", "sec")
        stkRoot.UnitPreferences.SetCurrentUnit("DateFormat", "EpSec")

        Dim scenario As IAgScenario = DirectCast(stkRoot.CurrentScenario, IAgScenario)
        scenario.Animation.AnimStepValue = 5

        '--Enable Next--
        CreateSatellite.Enabled = True
        '--Disable Current--
        CreateScenario.Enabled = False
    End Sub

    Private Sub CreateSatellite_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles CreateSatellite.Click
        '--Set Instructions--
        MainLabel.Text = My.Resources.Strings.CreateVector

        '--Create and Propogate m_satellite--
        m_satellite = DirectCast(stkRoot.CurrentScenario.Children.[New](AGI.STKObjects.AgESTKObjectType.eSatellite, "AGILE_31135"), AGI.STKObjects.AgSatellite)
        m_satellite.SetPropagatorType(AGI.STKObjects.AgEVePropagatorType.ePropagatorSGP4)
        Dim prop As AGI.STKObjects.IAgVePropagatorSGP4 = DirectCast(m_satellite.Propagator, AGI.STKObjects.IAgVePropagatorSGP4)
        prop.[Step] = 10
        prop.Propagate()

        '--Create Facility--
        m_facility = DirectCast(stkRoot.CurrentScenario.Children.[New](AgESTKObjectType.eFacility, "accessFac"), IAgFacility)
        m_facility.Graphics.LabelVisible = True
        m_facility.VO.Model.Visible = True

        '--Initialize m_satellite--
        m_satellite = DirectCast(stkRoot.CurrentScenario.Children("AGILE_31135"), AgSatellite)

        '--Set vector/angle Scale for appearance--
        m_satellite.VO.Vector.VectorSizeScale = 1.3
        m_satellite.VO.Vector.AngleSizeScale = 1

        '--Initialize _provider as IAgStkObject to access VGT--
        Dim provider As IAgCrdnProvider = m_satellite.Vgt

        '--Set up view point to use with "ZoomTo" function--
        '--Point coordinates relative to provider (m_satellite)--
        Dim viewPoint As IAgCrdnPointFixedInSystem = DirectCast(provider.Points.Factory.Create("ViewPoint", "View sat from this point", AgECrdnPointType.eCrdnPointTypeFixedInSystem), IAgCrdnPointFixedInSystem)
        viewPoint.FixedPoint.AssignCartesian(0.09, -0.1, -0.025)

        '--Enable Next--
        CreateVector.Enabled = True
        '--Disable current--
        CreateSatellite.Enabled = False
        '--Zoom to Satellite--
        ZoomTo()
    End Sub

    Private Sub CreateVector_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles CreateVector.Click
        '--Set Instructions--
        MainLabel.Text = My.Resources.Strings.ShowVelocityVector

        Dim provider As IAgCrdnProvider = m_satellite.Vgt

        stkRoot.BeginUpdate()
        Try
            '--Create Vector to accessFac--
            Dim vDisplacement As IAgCrdnVectorDisplacement = DirectCast(provider.Vectors.Factory.Create(m_sDistanceVectorName, "Vector to facility", AGI.STKVgt.AgECrdnVectorType.eCrdnVectorTypeDisplacement), IAgCrdnVectorDisplacement)
            vDisplacement.Origin.SetPoint(provider.Points("Center"))
            vDisplacement.Destination.SetPoint(TryCast(m_facility, IAgStkObject).Vgt.Points("Center"))
            vDisplacement.Apparent = True

            '--Show the Vectors--
            m_satellite.VO.Vector.RefCrdns.Add(AgEGeometricElemType.eVectorElem, TryCast(vDisplacement, IAgCrdn).QualifiedPath)
            Dim voVectorToFac As IAgVORefCrdnVector = DirectCast(m_satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.eVectorElem, TryCast(vDisplacement, IAgCrdn).QualifiedPath), IAgVORefCrdnVector)
            voVectorToFac.MagnitudeVisible = True
            voVectorToFac.LabelVisible = True
            voVectorToFac.Color = Color.AliceBlue
        Finally
            stkRoot.EndUpdate()
        End Try
        'Enable Next
        ShowVelocity.Enabled = True
        'Disable Current
        CreateVector.Enabled = False
    End Sub

    Private Sub ShowVelocity_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ShowVelocity.Click
        MainLabel.Text = My.Resources.Strings.CreateAxes

        Dim provider As IAgCrdnProvider = m_satellite.Vgt

        Dim path As String = TryCast(provider.Vectors("Velocity"), IAgCrdn).QualifiedPath

        stkRoot.BeginUpdate()
        Try
            '--Add Predefined Velocity Vector--
            Dim voVelocity As IAgVORefCrdnVector = DirectCast(m_satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.eVectorElem, path), IAgVORefCrdnVector)
            voVelocity.Visible = True
            voVelocity.MagnitudeVisible = True
            voVelocity.LabelVisible = True
        Finally
            stkRoot.EndUpdate()
        End Try
        '--Disable current--
        ShowVelocity.Enabled = False
        '--Enable next--
        CreateAxes.Enabled = True
    End Sub

    Private Sub CreateAxes_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles CreateAxes.Click
        MainLabel.Text = My.Resources.Strings.CreatePlane

        Dim provider As IAgCrdnProvider = m_satellite.Vgt

        Dim path As String = TryCast(provider.Axes("ICR"), IAgCrdn).QualifiedPath

        stkRoot.BeginUpdate()
        Try
            '--Add Predefined Axis--
            m_satellite.VO.Vector.RefCrdns.Add(AgEGeometricElemType.eAxesElem, path)
            Dim voAxes As IAgVORefCrdnAxes = DirectCast(m_satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.eAxesElem, path), IAgVORefCrdnAxes)
            voAxes.Visible = True
            voAxes.Color = Color.Tomato

            '--Hide velocity label--
            m_satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.eVectorElem, TryCast(provider.Vectors("Velocity"), IAgCrdn).QualifiedPath).Visible = False
        Finally
            stkRoot.EndUpdate()
        End Try
        '--Show next--
        CreatePlane.Enabled = True
        '--Disable Current--
        CreateAxes.Enabled = False
    End Sub

    Private Sub CreatePlane_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles CreatePlane.Click
        MainLabel.Text = My.Resources.Strings.CreateAngles

        Dim provider As IAgCrdnProvider = m_satellite.Vgt

        stkRoot.BeginUpdate()
        Try
            '--Create Plane Normal to Velocity Vector--
            Dim plane As IAgCrdnPlaneNormal = DirectCast(provider.Planes.Factory.Create("NormalPlane", "Normal to Velocity vector.", AgECrdnPlaneType.eCrdnPlaneTypeNormal), IAgCrdnPlaneNormal)
            plane.NormalVector.SetVector(provider.Vectors("Velocity"))
            plane.ReferenceVector.SetVector(provider.Vectors("Earth"))
            plane.ReferencePoint.SetPoint(provider.Points("Center"))

            Dim path As String = TryCast(provider.Planes("NormalPlane"), IAgCrdn).QualifiedPath
            '--Add plane to View--
            m_satellite.VO.Vector.RefCrdns.Add(AgEGeometricElemType.ePlaneElem, path)
            Dim voPlane As IAgVORefCrdnPlane = DirectCast(m_satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.ePlaneElem, path), IAgVORefCrdnPlane)
            voPlane.TransparentPlaneVisible = True
            voPlane.Color = Color.Yellow
            voPlane.AxisLabelsVisible = True

            '--Add BodyXY Plane--
            Dim pathXY As String = TryCast(provider.Planes("BodyXY"), IAgCrdn).QualifiedPath
            Dim voPlaneBodyXY As IAgVORefCrdnPlane = DirectCast(m_satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.ePlaneElem, pathXY), IAgVORefCrdnPlane)
            voPlaneBodyXY.Visible = True
            voPlaneBodyXY.TransparentPlaneVisible = True
            voPlaneBodyXY.Color = Color.Yellow
            voPlaneBodyXY.AxisLabelsVisible = True

            '--Hide Axes Labels--
            path = TryCast(provider.Axes("ICR"), IAgCrdn).QualifiedPath
            Dim axes As IAgVORefCrdnAxes = DirectCast(m_satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.eAxesElem, path), IAgVORefCrdnAxes)
            axes.LabelVisible = False
        Finally
            stkRoot.EndUpdate()
        End Try
        '--Show next--
        CreateAngle.Enabled = True
        '--Disable Current--
        CreatePlane.Enabled = False
    End Sub

    Private Sub CreateAngle_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles CreateAngle.Click
        '--Set Instructions--
        MainLabel.Text = My.Resources.Strings.Finish
        Dim vector3d As IAgVOVector = m_satellite.VO.Vector
        Dim provider As IAgCrdnProvider = m_satellite.Vgt

        stkRoot.BeginUpdate()
        Try
            '--Angle from Normal Plane to XY Plane (trajectory)--   
            Dim anglePlane As IAgCrdnAngleBetweenPlanes = DirectCast(provider.Angles.Factory.Create("AngleBetweenPlanes", "Angle from PlaneXY to NormalPlane", AgECrdnAngleType.eCrdnAngleTypeBetweenPlanes), IAgCrdnAngleBetweenPlanes)
            anglePlane.FromPlane.SetPlane(provider.Planes("BodyXY"))
            anglePlane.ToPlane.SetPlane(provider.Planes("NormalPlane"))

            '--Add to View--
            vector3d.RefCrdns.Add(AgEGeometricElemType.eAngleElem, DirectCast(anglePlane, IAgCrdn).QualifiedPath)
            Dim voAnglePlane As IAgVORefCrdnAngle = DirectCast(vector3d.RefCrdns.GetCrdnByName(AgEGeometricElemType.eAngleElem, DirectCast(anglePlane, IAgCrdn).QualifiedPath), IAgVORefCrdnAngle)
            voAnglePlane.AngleValueVisible = True

            '--Angle from Velocity/Trajector vector to AccessFacility vector--
            Dim angleVector As IAgCrdnAngleBetweenVectors = DirectCast(provider.Angles.Factory.Create("AngleToVector", "Angle from Vector to Plane", AgECrdnAngleType.eCrdnAngleTypeBetweenVectors), IAgCrdnAngleBetweenVectors)
            angleVector.FromVector.SetVector(provider.Vectors("Velocity"))
            angleVector.ToVector.SetVector(provider.Vectors(m_sDistanceVectorName))

            '--Add to View--
            vector3d.RefCrdns.Add(AgEGeometricElemType.eAngleElem, DirectCast(angleVector, IAgCrdn).QualifiedPath)
            Dim voAngleVector As IAgVORefCrdnAngle = DirectCast(vector3d.RefCrdns.GetCrdnByName(AgEGeometricElemType.eAngleElem, DirectCast(angleVector, IAgCrdn).QualifiedPath), IAgVORefCrdnAngle)
            voAngleVector.AngleValueVisible = True
            voAngleVector.Color = Color.SpringGreen

            '--Angle from AccessFacility vector to Plane--
            Dim anglePlaneFac As IAgCrdnAngleToPlane = DirectCast(provider.Angles.Factory.Create("AngleFromFac", "Angle from Plane to Vector", AgECrdnAngleType.eCrdnAngleTypeToPlane), IAgCrdnAngleToPlane)
            anglePlaneFac.ReferenceVector.SetVector(provider.Vectors(m_sDistanceVectorName))
            anglePlaneFac.ReferencePlane.SetPlane(provider.Planes("NormalPlane"))
            '--Add to View--
            m_satellite.VO.Vector.RefCrdns.Add(AgEGeometricElemType.eAngleElem, TryCast(anglePlaneFac, IAgCrdn).QualifiedPath)
            Dim voAngleFac As IAgVORefCrdnAngle = DirectCast(vector3d.RefCrdns.GetCrdnByName(AgEGeometricElemType.eAngleElem, TryCast(anglePlaneFac, IAgCrdn).QualifiedPath), IAgVORefCrdnAngle)
            voAngleFac.AngleValueVisible = True
        Finally
            stkRoot.EndUpdate()
        End Try
        '--Show next--
        CloseScenario1.Enabled = True
        '--Disable current--
        CreateAngle.Enabled = False
    End Sub

    Private Sub CloseScenario_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles CloseScenario1.Click
        CloseScenario()
        Application.[Exit]()
    End Sub

    Private Sub Button_ResetAnim_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button_ResetAnim.Click
        stkRoot.Rewind()
    End Sub

    Private Sub Button_StepBack_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button_StepBack.Click
        stkRoot.StepBackward()
    End Sub

    Private Sub Button_ReverseAnim_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button_ReverseAnim.Click
        stkRoot.PlayBackward()
    End Sub

    Private Sub Button_Pause_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button_Pause.Click
        stkRoot.Pause()
    End Sub

    Private Sub Button_Play_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button_Play.Click
        stkRoot.PlayForward()
    End Sub

    Private Sub Button_StepAhead_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button_StepAhead.Click
        stkRoot.StepForward()
    End Sub

    Private Sub Button_SlowDown_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button_SlowDown.Click
        stkRoot.Slower()
    End Sub

    Private Sub Button_SpeedUp_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button_SpeedUp.Click
        stkRoot.Faster()
    End Sub
    Private Sub ZoomTo()
        Dim viewPoint As IAgCrdnPointFixedInSystem = DirectCast(m_satellite.Vgt.Points(m_sViewPointName), IAgCrdnPointFixedInSystem)
        '--Zoom to m_satellite--
        Dim zoomCmd As String = String.Format("VO * View FromTo FromRegName ""STK Object"" FromName ""{1}"" ToRegName ""STK Object"" ToName ""{0}""", String.Format("{0}/{1}", m_satellite.ClassName, m_satellite.InstanceName), DirectCast(viewPoint, IAgCrdn).QualifiedPath)
        stkRootObject.ExecuteCommand(zoomCmd)
    End Sub
    Private Sub Unload(ByVal sender As Object, ByVal e As EventArgs)
        CloseScenario()
        Application.[Exit]()
    End Sub
End Class

File: CustomApplications/CSharp/VGTTutorial/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Data.Common;
using System.Collections;
using System.Collections.Specialized;
using System.Resources;
using System.Reflection;

using AGI.STKUtil;
using AGI.STKX;
using AGI.STKObjects;
using AGI.STKVgt;


namespace VGTTutorial
{
    public partial class VGT_Tutorial : Form
    {
        #region Private Data Members
        AgStkObjectRoot stkRootObject = null;
        AgSatellite _satellite = null;
        IAgFacility _facility = null;
        const string sViewPointName = "ViewPoint";
        const string sDistanceVectorName = "accessFac";
        ResourceManager resmgr = new ResourceManager("VGTTutorial.Strings", Assembly.GetExecutingAssembly());
        #endregion

        private AGI.STKObjects.AgStkObjectRoot stkRoot
        {
            get
            {
                if (stkRootObject == null)
                {
                    stkRootObject = new AGI.STKObjects.AgStkObjectRoot();
                    stkRootObject.OnAnimUpdate += new IAgStkObjectRootEvents_OnAnimUpdateEventHandler(stkRootObject_OnAnimUpdate);
                }
                return stkRootObject;
            }
        }

        void stkRootObject_OnAnimUpdate(double TimeEpSec)
        {
            if (_satellite != null)
            {
                IAgCrdnProvider provider = _satellite.Vgt;
                if (provider.Vectors.Contains(sDistanceVectorName))
                {
                    IAgCrdnVector vector = provider.Vectors[sDistanceVectorName];
                    IAgCrdnVectorFindInAxesResult result = vector.FindInAxes(TimeEpSec, provider.WellKnownAxes.Earth.Fixed);
                    if (result.IsValid)
                    {
                        // Calculate the vector's L2-norm.
                        IAgCartesian3Vector cartVec = result.Vector;
                        double mag = Math.Sqrt(
                            cartVec.X*cartVec.X + 
                            cartVec.Y*cartVec.Y +
                            cartVec.Z*cartVec.Z);

                        txtDistanceFromSatToFacility.Text = string.Format("{0} km", (mag/1000).ToString("F"));
                    }
                }
            }
        }

        void CloseScenario()
        {
            if (stkRootObject != null)
            {
                stkRootObject.CloseScenario();
            }
        }

        public VGT_Tutorial()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            MainLabel.Text = resmgr.GetString("Welcome");
        }

        private void NewScenario(object sender, EventArgs e)
        {
            //--Set Instructions--//
            MainLabel.Text = resmgr.GetString("CreateSatellite");
           
            //--Creates new scenario--//
            stkRoot.NewScenario("VGTTutorial");
            stkRoot.UnitPreferences.SetCurrentUnit("TimeUnit", "sec");
            stkRoot.UnitPreferences.SetCurrentUnit("DateFormat", "EpSec");

            IAgScenario scenario = stkRoot.CurrentScenario as IAgScenario;
            scenario.Animation.AnimStepValue = 5;

            //--Enable Next--//
            Create_Satellite.Enabled = true;
            //--Disable Current--//
            Create_Scenario.Enabled = false;
        }

        private void CreateSatellite(object sender, EventArgs e)
        {
            //--Set Instructions--//
            MainLabel.Text = resmgr.GetString("CreateVector");

            //--Create and Propogate _satellite--//
            _satellite = (AGI.STKObjects.AgSatellite)stkRoot.CurrentScenario.Children.New(
                         AGI.STKObjects.AgESTKObjectType.eSatellite, "AGILE_31135");
            _satellite.SetPropagatorType(AGI.STKObjects.AgEVePropagatorType.ePropagatorSGP4);
            AGI.STKObjects.IAgVePropagatorSGP4 prop = (AGI.STKObjects.IAgVePropagatorSGP4)_satellite.Propagator;
            prop.Step = 10;
            prop.Propagate();
     
            //--Create Facility--//
            _facility = (IAgFacility)stkRoot.CurrentScenario.Children.New(AgESTKObjectType.eFacility, "accessFac");
            _facility.Graphics.LabelVisible = true;
            _facility.VO.Model.Visible = true;
            
            //--Initialize _satellite--//
            _satellite = (AgSatellite)stkRoot.CurrentScenario.Children["AGILE_31135"];
            
            //--Set vector/angle Scale for appearance--//
            _satellite.VO.Vector.VectorSizeScale = 1.3;
            _satellite.VO.Vector.AngleSizeScale = 1;

            //--Initialize _provider as IAgStkObject to access VGT--//
            IAgCrdnProvider provider = _satellite.Vgt;
            
            //--Set up view point to use with "ZoomTo" function--//
            //--Point coordinates relative to provider (_satellite)--//
            IAgCrdnPointFixedInSystem viewPoint = (IAgCrdnPointFixedInSystem)provider.Points.Factory.Create(
                "ViewPoint", "View sat from this point", AgECrdnPointType.eCrdnPointTypeFixedInSystem);
            viewPoint.FixedPoint.AssignCartesian(.09, -.1, -.025);

            //--Enable Next--//
            Create_Vector.Enabled = true;
            //--Disable current--//
            Create_Satellite.Enabled = false;

            //--Zoom to Satellite--//
            ZoomTo();
        }

        private void CreateVector(object sender, EventArgs e)
        {
            //--Set Instructions--//
            MainLabel.Text = resmgr.GetString("ShowVelocityVector");

            IAgCrdnProvider provider = _satellite.Vgt;
            stkRoot.BeginUpdate();
            try
            {
                //--Create Vector to accessFac--//
                IAgCrdnVectorDisplacement vDisplacement = (IAgCrdnVectorDisplacement)provider.Vectors.Factory.Create(
                    sDistanceVectorName, "Vector to facility", AGI.STKVgt.AgECrdnVectorType.eCrdnVectorTypeDisplacement);
                vDisplacement.Origin.SetPoint(provider.Points["Center"]);
                vDisplacement.Destination.SetPoint((_facility as IAgStkObject).Vgt.Points["Center"]);
                vDisplacement.Apparent = true;

                //--Show the Vectors--//
                _satellite.VO.Vector.RefCrdns.Add(AgEGeometricElemType.eVectorElem, (vDisplacement as IAgCrdn).QualifiedPath);
                IAgVORefCrdnVector voVectorToFac = (IAgVORefCrdnVector)_satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.eVectorElem, (vDisplacement as IAgCrdn).QualifiedPath);
                voVectorToFac.MagnitudeVisible = true;
                voVectorToFac.LabelVisible = true;
                voVectorToFac.Color = Color.AliceBlue;
            }
            finally
            {
                stkRoot.EndUpdate();
            }
            //Enable Next//
            Show_Velocity_Vector.Enabled = true;
            //Disable Current//
            Create_Vector.Enabled = false;
        }

        private void ShowVelocityVector(object sender, EventArgs e)
        {
            MainLabel.Text = resmgr.GetString("CreateAxes");

            IAgCrdnProvider provider = _satellite.Vgt;

            string path = (provider.Vectors["Velocity"] as IAgCrdn).QualifiedPath;

            stkRoot.BeginUpdate();
            try
            {
                //--Show Predefined Velocity Vector--//
                IAgVORefCrdnVector voVelocity = (IAgVORefCrdnVector)_satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.eVectorElem, path);
                voVelocity.Visible = true;
                voVelocity.MagnitudeVisible = true;
                voVelocity.LabelVisible = true;
            }
            finally
            {
                stkRoot.EndUpdate();
            }
            //--Disable current--//
            Show_Velocity_Vector.Enabled = false;
            //--Enable next--//
            Create_Axes.Enabled = true;
        }      
    
        private void CreateAxes(object sender, EventArgs e)
        {
            MainLabel.Text = resmgr.GetString("CreatePlane");

            IAgCrdnProvider provider = _satellite.Vgt;
            string path = (provider.Axes["ICR"] as IAgCrdn).QualifiedPath;
            stkRoot.BeginUpdate();
            try
            {
                //--Add Predefined Axis--//
                _satellite.VO.Vector.RefCrdns.Add(AgEGeometricElemType.eAxesElem, path);
                IAgVORefCrdnAxes voAxes = (IAgVORefCrdnAxes)_satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.eAxesElem, path);
                voAxes.Visible = true;
                voAxes.Color = Color.Tomato;

                //--Hide velocity label--//
                _satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.eVectorElem,
                    (provider.Vectors["Velocity"] as IAgCrdn).QualifiedPath).Visible = false;
            }
            finally
            {
                stkRoot.EndUpdate();
            }
            //--Show next--//
            Create_Plane.Enabled = true;
            //--Disable Current--//
            Create_Axes.Enabled = false;
        }

        private void CreatePlane(object sender, EventArgs e)
        {
            MainLabel.Text = resmgr.GetString("CreateAngles");

            IAgCrdnProvider provider = _satellite.Vgt;

            stkRoot.BeginUpdate();
            try
            {
                //--Create Plane Normal to Velocity Vector--//
                IAgCrdnPlaneNormal plane = (IAgCrdnPlaneNormal)provider.Planes.Factory.Create("NormalPlane", "Normal to Velocity vector.",
                    AgECrdnPlaneType.eCrdnPlaneTypeNormal);
                plane.NormalVector.SetVector(provider.Vectors["Velocity"]);
                plane.ReferenceVector.SetVector(provider.Vectors["Earth"]);
                plane.ReferencePoint.SetPoint(provider.Points["Center"]);

                string path = (provider.Planes["NormalPlane"] as IAgCrdn).QualifiedPath;
                //--Add plane to View--//
                _satellite.VO.Vector.RefCrdns.Add(AgEGeometricElemType.ePlaneElem, path);
                IAgVORefCrdnPlane voPlane = (IAgVORefCrdnPlane)_satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.ePlaneElem, path);
                voPlane.TransparentPlaneVisible = true;
                voPlane.Color = Color.Yellow;
                voPlane.AxisLabelsVisible = true;

                //--Add BodyXY Plane--//
                string pathXY = (provider.Planes["BodyXY"] as IAgCrdn).QualifiedPath;
                IAgVORefCrdnPlane voPlaneBodyXY = (IAgVORefCrdnPlane)_satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.ePlaneElem, pathXY);
                voPlaneBodyXY.Visible = true;
                voPlaneBodyXY.TransparentPlaneVisible = true;
                voPlaneBodyXY.Color = Color.Yellow;
                voPlaneBodyXY.AxisLabelsVisible = true;

                //--Hide Axes Labels--//
                path = (provider.Axes["ICR"] as IAgCrdn).QualifiedPath;
                IAgVORefCrdnAxes axes = (IAgVORefCrdnAxes)_satellite.VO.Vector.RefCrdns.GetCrdnByName(AgEGeometricElemType.eAxesElem, path);
                axes.LabelVisible = false;
            }
            finally
            {
                stkRoot.EndUpdate();
            }
            //--Show next--//
            Create_Angle.Enabled = true;
            //--Disable Current--//
            Create_Plane.Enabled = false;
        }

        private void CreateAngle(object sender, EventArgs e)
        {
            //--Set Instructions--//
            MainLabel.Text = resmgr.GetString("Finish");

            IAgVOVector vector3d = _satellite.VO.Vector;
            IAgCrdnProvider provider = _satellite.Vgt;
            stkRoot.BeginUpdate();
            try
            {
                //--Angle from Normal Plane to XY Plane (trajectory)--//    
                IAgCrdnAngleBetweenPlanes anglePlane = (IAgCrdnAngleBetweenPlanes)provider.Angles.Factory.Create("AngleBetweenPlanes",
                    "Angle from PlaneXY to NormalPlane", AgECrdnAngleType.eCrdnAngleTypeBetweenPlanes);
                anglePlane.FromPlane.SetPlane(provider.Planes["BodyXY"]);
                anglePlane.ToPlane.SetPlane(provider.Planes["NormalPlane"]);

                //--Add to View--//
                vector3d.RefCrdns.Add(AgEGeometricElemType.eAngleElem, ((IAgCrdn)anglePlane).QualifiedPath);
                IAgVORefCrdnAngle voAnglePlane = (IAgVORefCrdnAngle)vector3d.RefCrdns.GetCrdnByName(
                    AgEGeometricElemType.eAngleElem, ((IAgCrdn)anglePlane).QualifiedPath);
                voAnglePlane.AngleValueVisible = true;

                //--Angle from Velocity/Trajector vector to AccessFacility vector--//
                IAgCrdnAngleBetweenVectors angleVector = (IAgCrdnAngleBetweenVectors)provider.Angles.Factory.Create("AngleToVector",
                    "Angle from Vector to Plane", AgECrdnAngleType.eCrdnAngleTypeBetweenVectors);
                angleVector.FromVector.SetVector(provider.Vectors["Velocity"]);
                angleVector.ToVector.SetVector(provider.Vectors[sDistanceVectorName]);

                //--Add to View--//
                vector3d.RefCrdns.Add(AgEGeometricElemType.eAngleElem, ((IAgCrdn)angleVector).QualifiedPath);
                IAgVORefCrdnAngle voAngleVector = (IAgVORefCrdnAngle)vector3d.RefCrdns.GetCrdnByName(
                    AgEGeometricElemType.eAngleElem, ((IAgCrdn)angleVector).QualifiedPath);
                voAngleVector.AngleValueVisible = true;
                voAngleVector.Color = Color.SpringGreen;

                //--Angle from AccessFacility vector to Plane--//
                IAgCrdnAngleToPlane anglePlaneFac = (IAgCrdnAngleToPlane)provider.Angles.Factory.Create("AngleFromFac",
                    "Angle from Plane to Vector", AgECrdnAngleType.eCrdnAngleTypeToPlane);
                anglePlaneFac.ReferenceVector.SetVector(provider.Vectors[sDistanceVectorName]);
                anglePlaneFac.ReferencePlane.SetPlane(provider.Planes["NormalPlane"]);
                //--Add to View--//
                _satellite.VO.Vector.RefCrdns.Add(AgEGeometricElemType.eAngleElem, (anglePlaneFac as IAgCrdn).QualifiedPath);
                IAgVORefCrdnAngle voAngleFac = (IAgVORefCrdnAngle)vector3d.RefCrdns.GetCrdnByName(
                    AgEGeometricElemType.eAngleElem, (anglePlaneFac as IAgCrdn).QualifiedPath);
                voAngleFac.AngleValueVisible = true;
            }
            finally
            {
                stkRoot.EndUpdate();
            }
            //--Show next--//
            Close_Scenario.Enabled = true;
            //--Disable current--//
            Create_Angle.Enabled = false;
        }

        private void Unload(object sender, EventArgs e)
        {
            CloseScenario();
            Application.Exit();
        }

        private void resetAnimButton_Click(object sender, EventArgs e)
        {
            stkRoot.Rewind();
        }

        private void startFwdAnimButton_Click(object sender, EventArgs e)
        {
            stkRoot.PlayForward();
        }

        private void pauseAnimButton_Click(object sender, EventArgs e)
        {
            stkRoot.Pause();
        }

        private void startRevAnimButton_Click(object sender, EventArgs e)
        {
            stkRoot.PlayBackward();
        }

        private void incTimeStepAnimButton_Click(object sender, EventArgs e)
        {
            stkRoot.Faster();
        }

        private void decTimeStepAnimButton_Click(object sender, EventArgs e)
        {
            stkRoot.Slower();
        }

        private void stepFwdAnimButton_Click(object sender, EventArgs e)
        {
            stkRoot.StepForward();
        }

        private void stepRevAnimButton_Click(object sender, EventArgs e)
        {
            stkRoot.StepBackward();
        }

        private void ZoomToSatClick(object sender, EventArgs e)
        {
            ZoomTo();
        }

        private void ZoomTo()
        {
            IAgCrdnPointFixedInSystem viewPoint = (IAgCrdnPointFixedInSystem)_satellite.Vgt.Points[sViewPointName];
            //--Zoom to _satellite--//
            string zoomCmd = string.Format("VO * View FromTo FromRegName \"STK Object\" FromName \"{1}\" ToRegName \"STK Object\" ToName \"{0}\"",
                string.Format("{0}/{1}",_satellite.ClassName, _satellite.InstanceName), 
                ((IAgCrdn)viewPoint).QualifiedPath);
            stkRootObject.ExecuteCommand(zoomCmd);
        }

        private void VGT_Tutorial_FormClosing(object sender, FormClosingEventArgs e)
        {
            CloseScenario();
        }
    }
}
